localization/imu_augmentor/src/imu_augmentor_nodelet.cc

```diff
@@ -0,0 +1,116 @@
+    : ff_util::FreeFlyerNodelet(NODE_IMU_AUG, true), platform_name_(GetPlatform()) {
+  imu_nh_.setCallbackQueue(&imu_queue_);
+  loc_nh_.setCallbackQueue(&loc_queue_);
+  heartbeat_.node = GetName();
+  heartbeat_.nodelet_manager = ros::this_node::getName();
+}
+
+void ImuAugmentorNodelet::Initialize(ros::NodeHandle* nh) {
+  SubscribeAndAdvertise(nh);
+  Run();
+}
+
+void ImuAugmentorNodelet::SubscribeAndAdvertise(ros::NodeHandle* nh) {
+  state_pub_ = nh->advertise<ff_msgs::EkfState>(TOPIC_GNC_EKF, 1);
+  pose_pub_ = nh->advertise<geometry_msgs::PoseStamped>(TOPIC_LOCALIZATION_POSE, 1);
+  twist_pub_ = nh->advertise<geometry_msgs::TwistStamped>(TOPIC_LOCALIZATION_TWIST, 1);
+  heartbeat_pub_ = nh->advertise<ff_msgs::Heartbeat>(TOPIC_HEARTBEAT, 5, true);
+
+  imu_sub_ = imu_nh_.subscribe(TOPIC_HARDWARE_IMU, 100, &ImuAugmentorNodelet::ImuCallback, this,
+                               ros::TransportHints().tcpNoDelay());
+  state_sub_ = loc_nh_.subscribe(TOPIC_GRAPH_LOC_STATE, 1, &ImuAugmentorNodelet::LocalizationStateCallback, this,
+                                 ros::TransportHints().tcpNoDelay());
+}
+
+void ImuAugmentorNodelet::ImuCallback(const sensor_msgs::Imu::ConstPtr& imu_msg) {
+  imu_augmentor_wrapper_.ImuCallback(*imu_msg);
+}
+
+void ImuAugmentorNodelet::LocalizationStateCallback(const ff_msgs::EkfState::ConstPtr& loc_msg) {
+  imu_augmentor_wrapper_.LocalizationStateCallback(*loc_msg);
+}
+
+boost::optional<ff_msgs::EkfState> ImuAugmentorNodelet::PublishLatestImuAugmentedLocalizationState() {
+  const auto latest_imu_augmented_loc_msg = imu_augmentor_wrapper_.LatestImuAugmentedLocalizationMsg();
+  if (!latest_imu_augmented_loc_msg) {
+    LogWarningEveryN(100, "PublishLatestImuAugmentedLocalizationState: Failed to get latest imu augmented loc msg.");
+    return boost::none;
+  }
+  state_pub_.publish(*latest_imu_augmented_loc_msg);
+  return latest_imu_augmented_loc_msg;
+}
+
+void ImuAugmentorNodelet::PublishPoseAndTwistAndTransform(const ff_msgs::EkfState& loc_msg) {
+  // Publish pose
+  geometry_msgs::PoseStamped pose_msg;
+  pose_msg.header = loc_msg.header;
+  pose_msg.pose = loc_msg.pose;
+  pose_pub_.publish(pose_msg);
+
+  // Publish twist
+  geometry_msgs::TwistStamped twist_msg;
+  twist_msg.header = loc_msg.header;
+  twist_msg.twist.linear = loc_msg.velocity;
+  twist_msg.twist.angular = loc_msg.omega;
+  twist_pub_.publish(twist_msg);
+
+  // Publish TF
+  geometry_msgs::TransformStamped transform_msg;
+  transform_msg.header = loc_msg.header;
+  transform_msg.child_frame_id = platform_name_ + "body";
+  transform_msg.transform.translation.x = loc_msg.pose.position.x;
+  transform_msg.transform.translation.y = loc_msg.pose.position.y;
+  transform_msg.transform.translation.z = loc_msg.pose.position.z;
+  transform_msg.transform.rotation = loc_msg.pose.orientation;
+  transform_pub_.sendTransform(transform_msg);
+}
+
+void ImuAugmentorNodelet::PublishHeartbeat() {
+  heartbeat_.header.stamp = ros::Time::now();
```

Comment:
The freeflyer nodelet publishes the heartbeat for you so you should be able to get rid of this.

Reply:
Right, the ff_nodelet publishes the heartbeat but only if you use the normal behavior of synchronous or asynchronous  spinning, in this case I bypass that and run a separate run function in a while loop so that I can add msgs from two different callback queues in batches sequentially, but it also means I have to publish my own heartbeat